### PR TITLE
 Migrate CircleCI workflows to GitHub Actions (1/3)

### PR DESCRIPTION
**Pull request by AzfaarQureshi**
### What this PR does

This is PR 1/3 of migrating Cortex's CI from CircleCI to GitHub Actions: 

**[Part 1/3](https://github.com/open-o11y/cortex/pull/6)** 👈 
* create `test-build-deploy` workflow under `.github/workflows`
* add readme
* adding the 3 base jobs
   * `lint`
   * `test`
   * `build`
* these jobs are run async and share no dependencies with each other
* workflow is run when a commit is pushed to `master` or when any PR is opened. However CD jobs will only run on the master branch and will be skipped otherwise
* all jobs in following PRs are dependant on one or more of these jobs finishing

**[Part 2/3](https://github.com/open-o11y/cortex/pull/7)** 
* adding jobs dependant on `build`
    * `integration`
    * `integration-config-db`
* This PR completes the CI portion of migration. The following PR will finish the remaining CD portion of the migration

**[Part 3/3](https://github.com/open-o11y/cortex/pull/8)** 
* This PR completes the migration from CircleCI to GitHub Actions
   * `deploy_website` is dependant on `build, test`
   * `deploy` is dependant on `build, test, lint, integration, integration-config-db`
* CD jobs will only run on the master branch and will be otherwise skipped

### Which issue(s) this PR fixes:
Fixes https://github.com/cortexproject/cortex/issues/3274

### Checklist
 - [ ] Tests updated <- Does not Apply to Workflow
- [x] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: .github/workflows/test-build-deploy.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
+name: ci
+on:
+  push:
+    branches: [master]
+    tags:
+      - v[0-9]+.[0-9]+.[0-9]+** # Tag filters not as strict due to different regex system on Github Actions
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml:6]: As of Oct 2020, GitHub [does not support](https://github.community/t/using-regex-for-filtering/16427/2)regex for tag filtering. The old regex  `/^v[0-9]+(\.[0-9]+){2}(-.+|[^-.]*)$/` was approximated using the available GitHub [filter patterns](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet)

File: .github/workflows/test-build-deploy.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    runs-on: ubuntu-latest
+    container:
+      image: quay.io/cortexproject/build-image:update-golang-1.14.9-eb0c8d4d2
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v1
+      - name: Sym Link Expected Path to Workspace
+        run: |
+          mkdir -p /go/src/github.com/cortexproject/cortex
+          ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml:19]: A significant number of commands in the Makefile were hardcoded with an assumed file structure of the CI container. make sure previous commands don’t break a symlink was created instead from the hardcoded “expected” working directory `go/src/github.com/cortexproject/cortex` to the actual working directory `$GITHUB_WORKSPACE`.

File: .github/workflows/test-build-deploy.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+          mkdir /tmp/images
+          ln -s /tmp/images ./docker-images
+          make BUILD_IN_CONTAINER=false save-images
+      - name: Compressing Images
+        run: tar -zcvf images.tar.gz /tmp/images
+      - name: Cache Images
+        uses: actions/upload-artifact@v2
+        with:
+          name: Docker Images
+          path: ./images.tar.gz
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml]: As of Oct 2020 GitHub Actions does not persist between different jobs in the same workflow. Each job is run on a fresh [virtual environment](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/introduction-to-github-actions#runners). As such, we need to upload and download artifacts to share data between jobs. 


Docker Images are zipped before uploading as a workaround. The images contain characters that are illegal in the upload-artifact action.

File: .github/workflows/test-build-deploy.yml [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        - 9042:9042
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v1
+      - name: Get Dependencies
+        run: go get -v -t -d ./...
+      - name: Sym Link Expected Path to Workspace
+        run: |
+          mkdir -p /go/src/github.com/cortexproject/cortex
+          ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml:52]: A significant number of commands in the Makefile were hardcoded with an assumed file structure of the CI container. make sure previous commands don’t break a symlink was created instead from the hardcoded “expected” working directory `/go/src/github.com/cortexproject/cortex` to the actual working directory `$GITHUB_WORKSPACE`.

File: .github/workflows/test-build-deploy.yml [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+          mkdir /tmp/images
+          ln -s /tmp/images ./docker-images
+          make BUILD_IN_CONTAINER=false save-images
+      - name: Zip Images
+        run: tar -zcvf images.tar.gz /tmp/images
+      - name: Upload Images Artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: Docker Images
+          path: ./images.tar.gz
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml:117]: As of Oct 2020 GitHub Actions does not persist between different jobs in the same workflow. Each job is run on a fresh [virtual environment](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/introduction-to-github-actions#runners). As such, we need to upload and download artifacts to share data between jobs.

Docker Images are zipped before uploading as a workaround. The images contain characters that are illegal in the upload-artifact action.

**Review by alolita — APPROVED** [review]
lgtm #shipit

**Pull request closed without merging**